.pre-commit-config.yaml

```diff
@@ -855,7 +855,7 @@ repos:
         files: ^dev/breeze/.*$
         pass_filenames: false
         require_serial: true
-        additional_dependencies: ['click', 'rich>=12.4.4', 'pyyaml']
+        additional_dependencies: ['click', 'rich>=12.4.4', 'pyyaml', 'methodtools']
```

Comment:
I hope this one is acceptable, because I have no idea how to decorate something without import it first.

Reply:
Indeed. Really bad idea to use lru_cache on class methods. 

But... Do we need to add methodtools? 

Why don't we replace all those usages with @cached_property actually ?

The only problematic one is I think __hash__ in `io` but it could be rather easily fixed without adding methodtools:

```python
  @cached_property
  def _hash(self):
      return hash(str(self))

  def __hash__(self):
      return self._hash
```

Reply:
Or am I missing something :) ?

Reply:
In case of `__hash__` I do not even know how it work before

```python
import functools

class A:
    @functools.lru_cache
    def __hash__(self) -> int:
        return hash(str(self))

hash(A()) # RecursionError: maximum recursion depth exceeded while calling a Python object
```

With cached property I can't see any problem here, so it could be used here instead of `methodtools`

```python
import functools

class A:
    @functools.cached_property
    def _hash(self):
        return hash(str(self))

    def __hash__(self) -> int:
        return self._hash

s = set()
a = A()
for _ in range(100):
    s.add(a)
assert len(s) == 1
del a
del s

import gc
gc.collect()
a_instances = [o for o in gc.get_objects() if isinstance(o, A)]
assert len(a_instances) == 0, len(a_instances)
```

Reply:
> But... Do we need to add methodtools?

If we could resolve all 4 cases without it, why not? One case with google provider could be also resolved by cached_property, but I was too lazy for rewrite tests 🤣

Reply:
Well it was to soon to say that this is a solution, I've create a test for check hashing and got and error in all case
There is different reason for that: __slots__, weakref do not work well with this types, recursion error. 


```python
    ...
    def test_hashing(self):
        o = ObjectStoragePath(f"file:///tmp/{str(uuid.uuid4())}")
        s = set()
        for _ in range(10):
            s.add(o)
        assert len(s) == 1
```

`@functools.lru_cache`
---

```console
    def test_hashing(self):
        o = ObjectStoragePath(f"file:///tmp/{str(uuid.uuid4())}")
        s = set()
        for _ in range(10):
>           s.add(o)
E           RecursionError: maximum recursion depth exceeded while calling a Python object
```

`@methodtools.lru_cache(max_size=None)`
---

```cosnsole
    def test_hashing(self):
        o = ObjectStoragePath(f"file:///tmp/{str(uuid.uuid4())}")
        s = set()
        for _ in range(10):
>           s.add(o)
E           TypeError: unhashable type: 'ObjectStoragePath'
```

Use cached property with name defined in `__slots__`
---


```console
tests/io/test_path.py:None (tests/io/test_path.py)
test_path.py:31: in <module>
    from airflow.io.path import ObjectStoragePath
../../airflow/io/path.py:72: in <module>
    class ObjectStoragePath(CloudPath):
E   ValueError: '_hash' in __slots__ conflicts with class variable
```

Use cached property with name not defined in `__slots__`
---

```console
>           raise TypeError(msg) from None
E           TypeError: No '__dict__' attribute on 'ObjectStoragePath' instance to cache '_awesome_hash' property.
```

@bolkedebruin I have a couple questions:
1. Do we actually need to make ObjectStoragePath hashable?
2. Do we need to make hash cachable?

Reply:
Old school caching by use class attribute  https://github.com/apache/airflow/pull/37757/commits/35a38d8a7154a6dac8cae3569fc20fb8773ab5bf